### DISCUSSION TITLE
InputHitTest method behaves strange with the Xunit Test Runner, leading to a maybe bigger issue

### DISCUSSION THREAD

**Discussion by yankun** [AvaloniaUI/Avalonia#12565] [Q&A]
**Describe the bug**
I've a simple test case, using the InputHitTest method of a window. If this method is used in only one test, everything is fine. However as soon as I add a second test, with the very same code, this test will fail if run along the other. The test will only succeed if it is run lonely. This is due to the fact that InputHitTest will start to return always null.

Further more, to have the test succeed at all, I have to call Dispatcher.UIThread.RunJobs() first.

**To Reproduce**
Please check the example Repro [here](https://github.com/yankun/XUnitBugExample) and run the tests. A single test will succeed, run both tests and the second one will fail. Remove Dispatcher.UIThread.RunJobs() and the test will always fail.

**Expected behavior**
That InputHitTest should succeeds always and that the call to Dispatcher.UIThread.RunJobs() should not be necessary.

****Additional context**
I'm having huge issues with the XUnit package in my integration tests. I'm having a solution with hundreds of integration tests and they will fail for sure if run within the same session, but will succeed if being run isolated one by one. Being run all at once the tests will start throwing "Platform not supported" exceptions or "Method not implemented" exceptions in the Avalonia code base at some point, as if the Application was not properly initialised. So a few tests are green and the rest will fail.**

Tested Versions: Avalonia 11.0.0-11.0.3

**Comment by maxkatz6** [comment]
> That InputHitTest should succeeds always and that the call to Dispatcher.UIThread.RunJobs() should not be necessary.


Requiring to call `Dispatcher.UIThread.RunJobs` is by design, as that's how real applications work.
Note: hit test is computed over rendering visuals and not layout.

InputHitTest by itself doesn't recalculate the layout or rendering scene. Instead, it uses previous rendering scene of visuals to compute a hit test operation.

In real-life applications there always will be a delay between window opening and user input, a delay during which render timer would run a couple of times and render frames based on the rendering scene.

With unit tests though, there is no delay. And `window.Show()` by itself doesn't trigger a render timer.
Furthermore, if you modify the controls tree (add a button later in the test), rendering scene won't be updated until the next rendered frame.
So, `Dispatcher.UIThread.RunJobs` is necessary there.

**Reply by yankun** [reply]
Thanks for the reply, now I understand why Dispatcher.UIThread.RunJobs is necessary there, but the question remains: Why does the InputHitTest method works only in one test and not the other, if run in the same test session. Have you checked out my example project und run the tests?

**Comment by maxkatz6** [comment]
> Being run all at once the tests will start throwing "Platform not supported" exceptions or "Method not implemented" exceptions in the Avalonia code base at some point, as if the Application was not properly initialised. So a few tests are green and the rest will fail.

This seems like a completely different issue. Can you open another issue with more details on it?
